cunumeric/deferred.py

```diff
@@ -305,37 +305,278 @@ def get_scalar_array(self):
                     )
-                tuple_of_arrays += (k,)
+            if store.transformed:
+                # after store is transformed we need to to return a copy of
+                # the store since Copy operation can't be done on
+                # the store with transformation
+                rhs, store = self._copy_store(store)
         else:
-            assert isinstance(key, NumPyThunk)
+            assert isinstance(key, DeferredArray)
+            # the use case when index array ndim >input array ndim
+            if key.ndim > store.ndim:
+                diff = store.ndim - key.ndim
+                for i in range(diff):
+                    store = store.promote(i + 1, store.shape[0])
+
+            if store.transformed:
+                rhs, store = self._copy_store(store)
+
             # Handle the boolean array case
-            if key.dtype == bool:
-                if key.ndim != self.ndim:
-                    raise TypeError(
-                        "Boolean advanced indexing dimension mismatch"
+            if key.dtype == np.bool:
+                if key.shape == rhs.shape:
+                    out_dtype = rhs.dtype
+                    if is_set:
+                        N = rhs.ndim
+                        out_dtype = rhs.runtime.get_point_type(N)
+
+                    out = rhs.runtime.create_unbound_thunk(out_dtype)
+                    task = rhs.context.create_task(
+                        CuNumericOpCode.ADVANCED_INDEXING
+                    )
+                    task.add_output(out.base)
+                    task.add_input(rhs.base)
+                    task.add_input(key.base)
+                    task.add_scalar_arg(is_set, bool)
+                    task.add_alignment(rhs.base, key.base)
+                    task.add_broadcast(
+                        rhs.base, axes=tuple(range(1, len(rhs.shape)))
+                    )
+                    task.add_broadcast(
+                        key.base, axes=tuple(range(1, len(key.shape)))
                     )
-                # For boolean arrays do the non-zero operation to make
-                # them into a normal indexing array
-                tuple_of_arrays = key.nonzero()
+                    task.execute()
+                    return False, rhs, out
+                else:
+                    # FIXME: replace `nonzero` case with the task with
+                    # output regions when ND output regions are available
+                    tuple_of_arrays = key.nonzero()
+            elif key.ndim < store.ndim:
+                output_arr = rhs._zip_indices(start_index, (key,))
+                return True, rhs, output_arr
             else:
-                tuple_of_arrays = (key,)
-        if len(tuple_of_arrays) != self.ndim:
+                tuple_of_arrays = (rhs.runtime.to_deferred_array(key),)
+
+        if len(tuple_of_arrays) > rhs.ndim:
             raise TypeError("Advanced indexing dimension mismatch")
```

Comment:
I think this case is already covered in the check below.

Reply:
Yes!